### PR TITLE
Add "--workspace" to update command

### PR DESCRIPTION
**Pull request by chaaz**
My `--bin` project has CI which updates the version number in `Cargo.toml`, which it then commits. However, this means that any further cargo command (`build`, `test`, etc) will update the existing `Cargo.lock` file (updating the root version), causing some frustration for users. Furthermore, it breaks the `publish` command, which requires the repo to be current.

I've added a `sync-lockfile` command to simply update the root version in the `Cargo.lock` file to match the `Cargo.toml` in the same way that simple commands like `fetch` do. If no `Cargo.lock` file is present, and new one is generated based on the index.

This is a demo PR for Pre-RFC conversation at https://internals.rust-lang.org/t/pre-rfc-cargo-command-to-just-sync-lockfile/13119, but may become a real PR if it gets approval.

### CODE REVIEW

**Comment by ehuss** [comment]
Thanks for the PR!  The Cargo team talked about this, and we feel that this would probably be better to be part of `cargo update`. There is a pretty high bar to add a new top-level command, and this doesn't feel like it really needs a new command. Perhaps a `--workspace` flag to mirror other commands?  I'm not certain that it is 100% equivalent to pass workspace members to the avoid list, but I believe it should be roughly the same.

**Comment by chaaz** [comment]
> Thanks for the PR! The Cargo team talked about this, and we feel that this would probably be better to be part of `cargo update`. There is a pretty high bar to add a new top-level command, and this doesn't feel like it really needs a new command. Perhaps a `--workspace` flag to mirror other commands? I'm not certain that it is 100% equivalent to pass workspace members to the avoid list, but I believe it should be roughly the same.

Thanks, @ehuss. Discussion on the forums came to roughly the same conclusion. I'll change the PR so that `update --workspace` does what I want, instead of an entirely new command.

Are the Cargo Book descriptions auto-pulled from the source, or should I be updating that separately?

**Comment by ehuss** [comment]
> Are the Cargo Book descriptions auto-pulled from the source, or should I be updating that separately?

The documentation is written in markdown. The `cargo-update` documentation is here: https://github.com/rust-lang/cargo/blob/master/src/doc/man/cargo-update.md.  After updating the docs, you need to run `build-man.sh` to rebuild the man pages. There's more documentation on how this works at https://github.com/rust-lang/cargo/tree/master/src/doc.

**Comment by chaaz** [comment]
@ehuss Sorry for the slow response: real life intervened. Anywho, I have moved the functionality to `update --workspace` as discussed, and have added the flag to the docs as well. Please let me know how that looks.

File: src/bin/cargo/commands/update.rs [hunk 1]
```diff
@@ -7,6 +7,7 @@ pub fn cli() -> App {
     subcommand("update")
         .about("Update dependencies as recorded in the local lock file")
         .arg(opt("quiet", "No output printed to stdout").short("q"))
+        .arg(opt("workspace", "Only update the workspace pakcages").short("w"))
```
Comment (ehuss) [inline src/bin/cargo/commands/update.rs:10]:
```suggestion
        .arg(opt("workspace", "Only update the workspace packages").short("w"))
```

**Review by ehuss — COMMENTED** [review]
Thanks!

From an implementation standpoint, I was thinking that the `--workspace` flag would just add the set of member `PackageId`s to the `to_avoid` list.  That would avoid needing a special code path.  This current implementation seems to ignore additional `-p` flags, and doesn't report what was updated.  Do you think there would be any particular problems with that approach, or were there other reasons to go with the current design?

**Comment by chaaz** [comment]
> Thanks!
> 
> From an implementation standpoint, I was thinking that the `--workspace` flag would just add the set of member `PackageId`s to the `to_avoid` list. That would avoid needing a special code path. This current implementation seems to ignore additional `-p` flags, and doesn't report what was updated. Do you think there would be any particular problems with that approach, or were there other reasons to go with the current design?

Thanks, that does seem like a better approach. Will put that in.

**Comment by chaaz** [comment]
@ehuss Just checked in the suggested changes: looks like the best way is just to keep the `to_avoid` array empty (-ish) if there's a `--workspaces` flag. That is much easier approach, and integrates better with existing behavior. Let me know what you think.

**Comment by ehuss** [comment]
@rfcbot fcp merge

This adds a new stable flag `--workspace` (or `-w`) to `cargo update` which will cause it to update only the workspace members (or more specifically, it does an update equivalent to `cargo build` which updates any out-of-date path sources).  This can be useful when bumping the version in `Cargo.toml`, and you want to update `Cargo.lock`, but don't want to run some other command (like `build` or `metadata`), or don't want to explicitly list each workspace member like `cargo update -p foo`.

**Comment by ehuss** [comment]
@rfcbot fcp merge

**Pull request merged**